File: 2024/06/Node/main.js
const fs = require('fs');

const directions = [
  [-1, 0], 
  [0, 1], 
  [1, 0], 
  [0, -1]
];

function parseInput(filename) {
  const grid = fs.readFileSync(filename, 'utf8').split('\n');
  let guardPos = null;
  let guardDir = 0;

  for (let row = 0; row < grid.length; row++) {
    for (let col = 0; col < grid[row].length; col++) {
      const cell = grid[row][col];
      if (cell === '^') { guardPos = [row, col]; guardDir = 0; }
      if (cell === '>') { guardPos = [row, col]; guardDir = 1; }
      if (cell === 'v') { guardPos = [row, col]; guardDir = 2; }
      if (cell === '<') { guardPos = [row, col]; guardDir = 3; }
    }
  }

  return { grid, guardPos, guardDir };
}

function simulateGuardMovement(grid, guardPos, guardDir) {
  const visited = new Set();
  let visitedPositions = 0;

  const visitPosition = (pos) => {
    const posStr = `${pos[0]},${pos[1]}`;
    if (!visited.has(posStr)) {
      visited.add(posStr);
      visitedPositions++;
    }
  };

  visitPosition(guardPos);

  while (true) {
    const [x, y] = guardPos;
    const [dx, dy] = directions[guardDir];
    const newPos = [x + dx, y + dy];

    if (newPos[0] < 0 || newPos[0] >= grid.length || newPos[1] < 0 || newPos[1] >= grid[newPos[0]].length) break;

    if (grid[newPos[0]][newPos[1]] === '#') {
      guardDir = (guardDir + 1) % 4;
    } else {
      guardPos = newPos;
      visitPosition(guardPos);
    }
  }

  return visitedPositions;
}

function main() {
  const { grid, guardPos, guardDir } = parseInput('../input.txt');
  const distinctPositions = simulateGuardMovement(grid, guardPos, guardDir);
  console.log(`Distinct positions visited: ${distinctPositions}`);
}

main();

File: 2024/06/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var directions = [4][2]int{
	{-1, 0},
	{0, 1},
	{1, 0},
	{0, -1}, 
}

func parseInput(filename string) ([][]string, [2]int, int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, [2]int{}, 0, err
	}
	defer file.Close()

	var grid [][]string
	var guardPos [2]int
	var guardDir int

	scanner := bufio.NewScanner(file)
	row := 0
	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, strings.Split(line, ""))
		for col, cell := range grid[row] {
			switch cell {
			case "^":
				guardPos = [2]int{row, col}
				guardDir = 0
			case ">":
				guardPos = [2]int{row, col}
				guardDir = 1
			case "v":
				guardPos = [2]int{row, col}
				guardDir = 2
			case "<":
				guardPos = [2]int{row, col}
				guardDir = 3
			}
		}
		row++
	}

	if err := scanner.Err(); err != nil {
		return nil, [2]int{}, 0, err
	}

	return grid, guardPos, guardDir, nil
}

func simulateGuardMovement(grid [][]string, guardPos [2]int, guardDir int) int {
	visited := make(map[string]bool)
	visitedPositions := 0

	visitPosition := func(pos [2]int) {
		posStr := fmt.Sprintf("%d,%d", pos[0], pos[1])
		if !visited[posStr] {
			visited[posStr] = true
			visitedPositions++
		}
	}

	visitPosition(guardPos)

	for {
		x, y := guardPos[0], guardPos[1]
		dx, dy := directions[guardDir][0], directions[guardDir][1]
		newPos := [2]int{x + dx, y + dy}

		if newPos[0] < 0 || newPos[0] >= len(grid) || newPos[1] < 0 || newPos[1] >= len(grid[newPos[0]]) {
			break
		}
		if grid[newPos[0]][newPos[1]] == "#" {
			guardDir = (guardDir + 1) % 4
		} else {
			guardPos = newPos
			visitPosition(guardPos)
		}
	}

	return visitedPositions
}

func main() {
	grid, guardPos, guardDir, err := parseInput("../input.txt")
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	distinctPositions := simulateGuardMovement(grid, guardPos, guardDir)
	fmt.Printf("Distinct positions visited: %d\n", distinctPositions)
}
